main/src/ca/uwaterloo/flix/api/lsp/provider/HoverProvider.scala

```diff
@@ -28,56 +30,42 @@ import scala.annotation.tailrec
+  def processHover(uri: String, pos: Position)(implicit root: Root, flix: Flix): JObject = {
+    var stack: List[AnyRef] = Nil
+    def push(x: AnyRef): Unit = {
+      stack = x :: stack
+    }
+    object HoverConsumer extends Visitor.Consumer {
+      override def consumeType(tpe: Type): Unit = push(tpe)
+      override def consumeExpr(exp: Expr): Unit = push(exp)
+      override def consumeDefSymUse(symUse: DefSymUse): Unit = push(symUse)
+      override def consumeSigSymUse(symUse: SigSymUse): Unit = push(symUse)
+      override def consumeOpSymUse(symUse: OpSymUse): Unit = push(symUse)
+      override def consumeDef(defn: Def): Unit = push(defn)
+      override def consumeSig(sig: Sig): Unit = push(sig)
+      override def consumeOp(op: Op): Unit = push(op)
+      override def consumeFormalParam(fparam: FormalParam): Unit = push(fparam)
     }
-  }
-
-  @tailrec
-  private def hoverEntity(entity: Entity, uri: String, pos: Position)(implicit Index: Index, root: Root, flix: Flix): JObject = entity match {
-
-    case Entity.Case(caze) => hoverType(caze.tpe, caze.sym.loc)
-
-    case Entity.StructField(field) => hoverType(field.tpe, field.sym.loc)
-
-    case Entity.DefUse(sym, loc, _) => hoverDef(sym, loc)
-
-    case Entity.SigUse(sym, loc, _) => hoverSig(sym, loc)
-
-    case Entity.VarUse(_, _, parent) => hoverEntity(parent, uri, pos)
-
-    case Entity.CaseUse(_, _, parent) => hoverEntity(parent, uri, pos)
-
-    case Entity.StructFieldUse(_, _, parent) => hoverEntity(parent, uri, pos)
-
-    case Entity.Exp(exp) => hoverTypeAndEff(exp.tpe, exp.eff, exp.loc)
-
-    case Entity.FormalParam(fparam) => hoverType(fparam.tpe, fparam.loc)
-
-    case Entity.Pattern(pat) => hoverType(pat.tpe, pat.loc)
-
-    case Entity.Pred(pred, tpe) => hoverType(tpe, pred.loc)
-
-    case Entity.LocalVar(sym, tpe) => hoverType(tpe, sym.loc)
 
-    case Entity.Type(t) => hoverKind(t)
+    Visitor.visitRoot(root, HoverConsumer, Visitor.InsideAcceptor(uri, pos))
 
-    case Entity.OpUse(sym, loc, _) => hoverOp(sym, loc)
+    stack match {
+      case Nil => mkNotFound(uri, pos)
+      case head :: _ => hoverAny(head, uri, pos)
+    }
+  }
 
-    case Entity.Trait(_) => mkNotFound(uri, pos)
-    case Entity.Def(_) => mkNotFound(uri, pos)
-    case Entity.Effect(_) => mkNotFound(uri, pos)
-    case Entity.Enum(_) => mkNotFound(uri, pos)
-    case Entity.Struct(_) => mkNotFound(uri, pos)
-    case Entity.TypeAlias(_) => mkNotFound(uri, pos)
-    case Entity.AssocType(_) => mkNotFound(uri, pos)
-    case Entity.Label(_) => mkNotFound(uri, pos)
-    case Entity.Op(_) => mkNotFound(uri, pos)
-    case Entity.Sig(_) => mkNotFound(uri, pos)
-    case Entity.TypeVar(_) => mkNotFound(uri, pos)
+  private def hoverAny(x: AnyRef, uri: String, pos: Position)(implicit root: Root, flix: Flix): JObject = x match {
+    case tpe: Type => hoverKind(tpe)
+    case exp: Expr => hoverTypeAndEff(exp.tpe, exp.eff, exp.loc)
+    case DefSymUse(sym, loc) => hoverDef(sym, loc)
```

Comment:
What's the argument to support both Def and DefSymUse? I think that hover on a def, when it is right there, should not show the same information...

Reply:
I thought that would be expected. But I can see your point that it's redundant. Alternatively, howering on the definition itself, could do nothing or show like the arrow type of the function (although this is also somewhat redundant as it can easily be gleaned from the signature)

Reply:
It seems redundant to me.

Reply:
yep, Def, Sig and Op hover has been removed